src/rrd_create.c

```diff
@@ -1437,14 +1437,21 @@ int write_fh(
         /* we output 3 or higher */
         strcpy(rrd->stat_head->version, "0003");
     }
-    fwrite(rrd->stat_head, sizeof(stat_head_t), 1, fh);
-    fwrite(rrd->ds_def, sizeof(ds_def_t), rrd->stat_head->ds_cnt, fh);
-    fwrite(rrd->rra_def, sizeof(rra_def_t), rrd->stat_head->rra_cnt, fh);
-    fwrite(rrd->live_head, sizeof(live_head_t), 1, fh);
-    fwrite(rrd->pdp_prep, sizeof(pdp_prep_t), rrd->stat_head->ds_cnt, fh);
-    fwrite(rrd->cdp_prep, sizeof(cdp_prep_t),
+
+#define FWRITE_CHECK(ptr, size, nitems, fp)                    \
+    do {                                                       \
+        if (fwrite((ptr), (size), (nitems), (fp)) != (nitems)) \
+            return (-1);                                       \
+    } while (0)
```

Comment:
1) this is to protect against the following:

if (...)
   FWRITE_CHECK(...);
else
   /* do something */

Without do/while 'else' will be related to the 'if' inside the FWRITE_CHECK (and this is not what the user of this macro is expected).

2) It is generally a good practice to wrap macros that do not expand to expressions (like this one, or any multi-statement macros) so that the user of macro can use it as a single statement (as in the example above). do/while(0) will anyway be optimized out but it turns a multi-statement macro to a "single-statement".